File: solutions/02_agentic_coding/03_translate_script.cpp
/**
 * Solution 2.3: Translation of Python sequence utilities to C++17
 *
 * Compile with:
 *   g++ -std=c++17 -O2 -o sequence_utils 03_translate_script.cpp
 *
 * Run with:
 *   ./sequence_utils
 */

#include <algorithm>
#include <cmath>
#include <iostream>
#include <map>
#include <stdexcept>
#include <string>
#include <tuple>
#include <vector>

/**
 * @brief Calculate the GC content of a DNA sequence as a percentage.
 *
 * @param sequence DNA sequence string (A, T, G, C; case-insensitive).
 * @return GC content as a percentage (0.0 – 100.0).
 * @throws std::invalid_argument if the sequence contains non-DNA characters.
 */
double calculate_gc_content(const std::string& sequence) {
    if (sequence.empty()) {
        return 0.0;
    }

    std::string upper_seq = sequence;
    std::transform(upper_seq.begin(), upper_seq.end(), upper_seq.begin(), ::toupper);

    for (char c : upper_seq) {
        if (c != 'A' && c != 'C' && c != 'G' && c != 'T') {
            throw std::invalid_argument("Invalid nucleotide in sequence: " + sequence);
        }
    }

    int gc_count = 0;
    for (char c : upper_seq) {
        if (c == 'G' || c == 'C') {
            ++gc_count;
        }
    }
    return (static_cast<double>(gc_count) / static_cast<double>(upper_seq.size())) * 100.0;
}

/**
 * @brief Convert a DNA sequence to its RNA equivalent.
 *
 * Replaces thymine (T/t) with uracil (U/u), preserving original case.
 *
 * @param sequence DNA sequence string.
 * @return RNA sequence string.
 * @throws std::invalid_argument if the sequence contains non-DNA characters.
 */
std::string transcribe_dna_to_rna(const std::string& sequence) {
    std::string upper_seq = sequence;
    std::transform(upper_seq.begin(), upper_seq.end(), upper_seq.begin(), ::toupper);

    for (char c : upper_seq) {
        if (c != 'A' && c != 'C' && c != 'G' && c != 'T') {
            throw std::invalid_argument("Invalid nucleotide in sequence: " + sequence);
        }
    }

    std::string result = sequence;
    for (char& c : result) {
        if (c == 'T') c = 'U';
        else if (c == 't') c = 'u';
    }
    return result;
}

/**
 * @brief Return the reverse complement of a DNA sequence (uppercase output).
 *
 * @param sequence DNA sequence string (case-insensitive).
 * @return Reverse complement as an uppercase string.
 * @throws std::invalid_argument if the sequence contains non-DNA characters.
 */
std::string reverse_complement(const std::string& sequence) {
    std::string upper_seq = sequence;
    std::transform(upper_seq.begin(), upper_seq.end(), upper_seq.begin(), ::toupper);

    for (char c : upper_seq) {
        if (c != 'A' && c != 'C' && c != 'G' && c != 'T') {
            throw std::invalid_argument("Invalid nucleotide in sequence: " + sequence);
        }
    }

    std::string result;
    result.reserve(upper_seq.size());

    const std::map<char, char> complement = {{'A', 'T'}, {'T', 'A'}, {'G', 'C'}, {'C', 'G'}};
    for (auto it = upper_seq.rbegin(); it != upper_seq.rend(); ++it) {
        result += complement.at(*it);
    }
    return result;
}

/**
 * @brief Calculate the melting temperature (Tm) of a DNA primer.
 *
 * Uses the Wallace rule for primers up to 13 bp:
 *   Tm = 2*(A+T) + 4*(G+C)
 *
 * Uses a simplified nearest-neighbour approximation for longer primers:
 *   Tm = 81.5 + 16.6*log10([Na+]) + 0.41*%GC - 675/n
 * with [Na+] = 0.05 M.
 *
 * @param sequence DNA primer sequence (case-insensitive).
 * @return Melting temperature in degrees Celsius, rounded to 2 decimal places.
 */
double calculate_melting_temperature(const std::string& sequence) {
    std::string upper_seq = sequence;
    std::transform(upper_seq.begin(), upper_seq.end(), upper_seq.begin(), ::toupper);

    int n = static_cast<int>(upper_seq.size());
    int gc_count = 0;
    int at_count = 0;
    for (char c : upper_seq) {
        if (c == 'G' || c == 'C') ++gc_count;
        else if (c == 'A' || c == 'T') ++at_count;
    }

    double tm;
    if (n <= 13) {
        tm = 2.0 * at_count + 4.0 * gc_count;
    } else {
        double gc_percent = (static_cast<double>(gc_count) / n) * 100.0;
        double sodium_concentration = 0.05;
        tm = 81.5 + 16.6 * std::log10(sodium_concentration) + 0.41 * gc_percent - 675.0 / n;
    }

    // Round to 2 decimal places
    return std::round(tm * 100.0) / 100.0;
}

/**
 * @brief Find restriction enzyme recognition sites in a DNA sequence.
 *
 * @param sequence DNA sequence to search (case-insensitive).
 * @param enzyme_sites Map from enzyme name to recognition sequence.
 * @return Vector of (enzyme_name, position) pairs sorted by position.
 *         Positions are 0-indexed.
 */
std::vector<std::pair<std::string, std::size_t>>
find_restriction_sites(const std::string& sequence,
                        const std::map<std::string, std::string>& enzyme_sites) {
    std::string upper_seq = sequence;
    std::transform(upper_seq.begin(), upper_seq.end(), upper_seq.begin(), ::toupper);

    std::vector<std::pair<std::string, std::size_t>> results;

    for (const auto& [enzyme, site] : enzyme_sites) {
        std::string upper_site = site;
        std::transform(upper_site.begin(), upper_site.end(), upper_site.begin(), ::toupper);

        std::size_t pos = 0;
        while ((pos = upper_seq.find(upper_site, pos)) != std::string::npos) {
            results.emplace_back(enzyme, pos);
            ++pos;
        }
    }

    std::sort(results.begin(), results.end(),
              [](const auto& a, const auto& b) { return a.second < b.second; });

    return results;
}

// ── main ─────────────────────────────────────────────────────────────────────

int main() {
    const std::string test_sequence =
        "ATGCGAATTCGGATCCATGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCG";

    std::cout << "=== Sequence Analysis ===" << std::endl;
    std::cout << "Sequence: " << test_sequence << std::endl;
    std::cout << "GC content: " << calculate_gc_content(test_sequence) << "%" << std::endl;
    std::cout << "RNA transcript: " << transcribe_dna_to_rna(test_sequence) << std::endl;
    std::cout << "Reverse complement: " << reverse_complement(test_sequence) << std::endl;
    std::cout << "Melting temperature: " << calculate_melting_temperature(test_sequence)
              << " C" << std::endl;

    std::map<std::string, std::string> enzyme_sites = {
        {"EcoRI",   "GAATTC"},
        {"BamHI",   "GGATCC"},
        {"HindIII", "AAGCTT"},
    };

    auto sites = find_restriction_sites(test_sequence, enzyme_sites);
    std::cout << "\nRestriction sites found:" << std::endl;
    for (const auto& [enzyme, pos] : sites) {
        std::cout << "  " << enzyme << " at position " << pos << std::endl;
    }

    return 0;
}

File: solutions/02_agentic_coding/03_translate_script.R
# Solution 2.3: Translation of Python sequence utilities to R
# Using base R and stringr for string operations

library(stringr)

#' Calculate the GC content of a DNA sequence
#'
#' @param sequence A character string containing the DNA sequence (A, T, G, C).
#'   Both uppercase and lowercase are accepted.
#' @return A numeric value representing the GC content as a percentage (0–100).
#' @examples
#' calculate_gc_content("ATGCGC")  # returns 66.66667
#' calculate_gc_content("AAAA")    # returns 0
calculate_gc_content <- function(sequence) {
  sequence <- toupper(sequence)
  valid_bases <- c("A", "C", "G", "T")
  chars <- strsplit(sequence, "")[[1]]

  if (!all(chars %in% valid_bases)) {
    stop(paste("Invalid nucleotides in sequence:", sequence))
  }
  if (nchar(sequence) == 0) {
    return(0.0)
  }
  gc_count <- sum(chars %in% c("G", "C"))
  return((gc_count / nchar(sequence)) * 100.0)
}


#' Transcribe a DNA sequence to RNA
#'
#' Replaces all thymine (T/t) with uracil (U/u), preserving the original case.
#'
#' @param sequence A character string containing the DNA sequence.
#' @return A character string representing the RNA transcript.
#' @examples
#' transcribe_dna_to_rna("ATGCTT")  # returns "AUGCUU"
#' transcribe_dna_to_rna("atgctt")  # returns "augcuu"
transcribe_dna_to_rna <- function(sequence) {
  sequence_upper <- toupper(sequence)
  valid_bases <- c("A", "C", "G", "T")
  chars <- strsplit(sequence_upper, "")[[1]]

  if (!all(chars %in% valid_bases)) {
    stop(paste("Invalid nucleotides in sequence:", sequence))
  }
  rna <- chartr("Tt", "Uu", sequence)
  return(rna)
}


#' Return the reverse complement of a DNA sequence
#'
#' @param sequence A character string containing the DNA sequence.
#' @return A character string containing the reverse complement (uppercase).
#' @examples
#' reverse_complement("ATGC")          # returns "GCAT"
#' reverse_complement("AAAAACCCGGT")   # returns "ACCGGGTTTT"
reverse_complement <- function(sequence) {
  sequence_upper <- toupper(sequence)
  valid_bases <- c("A", "C", "G", "T")
  chars <- strsplit(sequence_upper, "")[[1]]

  if (!all(chars %in% valid_bases)) {
    stop(paste("Invalid nucleotides in sequence:", sequence))
  }
  complement_map <- c(A = "T", T = "A", G = "C", C = "G")
  complemented <- complement_map[chars]
  return(paste(rev(complemented), collapse = ""))
}


#' Calculate the melting temperature (Tm) of a DNA primer
#'
#' Uses the Wallace rule for short primers (<= 13 bp):
#' \deqn{T_m = 2 \cdot (A + T) + 4 \cdot (G + C)}
#' Uses a simplified nearest-neighbour approximation for longer primers.
#'
#' @param sequence A character string containing the primer sequence (DNA).
#' @return A numeric value representing the melting temperature in degrees Celsius.
#' @examples
#' calculate_melting_temperature("ATGCGC")  # short primer — uses Wallace rule
calculate_melting_temperature <- function(sequence) {
  sequence_upper <- toupper(sequence)
  chars <- strsplit(sequence_upper, "")[[1]]
  n <- length(chars)
  gc_count <- sum(chars %in% c("G", "C"))
  at_count <- sum(chars %in% c("A", "T"))

  if (n <= 13) {
    tm <- 2.0 * at_count + 4.0 * gc_count
  } else {
    gc_percent <- (gc_count / n) * 100.0
    sodium_concentration <- 0.05
    tm <- 81.5 + 16.6 * log10(sodium_concentration) + 0.41 * gc_percent - 675.0 / n
  }
  return(round(tm, 2))
}


#' Find restriction enzyme recognition sites in a DNA sequence
#'
#' @param sequence A character string containing the DNA sequence to search.
#' @param enzyme_sites A named character vector mapping enzyme names to their
#'   recognition sequences. Example:
#'   \code{c(EcoRI = "GAATTC", BamHI = "GGATCC")}
#' @return A data frame with columns \code{enzyme} and \code{position}
#'   (0-indexed), sorted by position.
#' @examples
#' sites <- find_restriction_sites(
#'   "ATGCGAATTCGGATCC",
#'   c(EcoRI = "GAATTC", BamHI = "GGATCC")
#' )
find_restriction_sites <- function(sequence, enzyme_sites) {
  sequence_upper <- toupper(sequence)
  results <- list()

  for (enzyme in names(enzyme_sites)) {
    site <- toupper(enzyme_sites[[enzyme]])
    # gregexpr returns 1-indexed positions; convert to 0-indexed
    matches <- gregexpr(site, sequence_upper, fixed = TRUE)[[1]]
    if (matches[1] != -1) {
      for (pos in matches) {
        results <- c(results, list(data.frame(enzyme = enzyme, position = pos - 1L)))
      }
    }
  }

  if (length(results) == 0) {
    return(data.frame(enzyme = character(0), position = integer(0)))
  }

  result_df <- do.call(rbind, results)
  result_df <- result_df[order(result_df$position), ]
  rownames(result_df) <- NULL
  return(result_df)
}


# --- Main ---
test_sequence <- "ATGCGAATTCGGATCCATGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCG"

cat("=== Sequence Analysis ===\n")
cat("Sequence:", test_sequence, "\n")
cat(sprintf("GC content: %.1f%%\n", calculate_gc_content(test_sequence)))
cat("RNA transcript:", transcribe_dna_to_rna(test_sequence), "\n")
cat("Reverse complement:", reverse_complement(test_sequence), "\n")
cat(sprintf("Melting temperature: %.2f °C\n", calculate_melting_temperature(test_sequence)))

enzyme_sites <- c(EcoRI = "GAATTC", BamHI = "GGATCC", HindIII = "AAGCTT")
sites <- find_restriction_sites(test_sequence, enzyme_sites)
cat("\nRestriction sites found:\n")
for (i in seq_len(nrow(sites))) {
  cat(sprintf("  %s at position %d\n", sites$enzyme[i], sites$position[i]))
}
